pkg/tcpip/stack/conntrack.go

```diff
@@ -0,0 +1,253 @@
+			tcpHeader = header.TCP(newPkt.TransportHeader)
+		} else {
+			if len(pkt.Data.First()) < header.TCPMinimumSize {
+				return tuple, tcpip.ErrUnknownProtocol
+			}
+			tcpHeader = header.TCP(newPkt.Data.First())
+		}
+	} else {
+		netHeader = header.IPv4(pkt.NetworkHeader)
+		// TODO(gvisor.dev/issue/170): Need to support for other
+		// protocols as well.
+		if netHeader == nil || netHeader.TransportProtocol() != header.TCPProtocolNumber {
+			return tuple, tcpip.ErrUnknownProtocol
+		}
+		tcpHeader = header.TCP(pkt.TransportHeader)
+	}
+
+	tuple.src.addr = netHeader.SourceAddress()
+	tuple.src.port = tcpHeader.SourcePort()
+	tuple.src.protocol = uint16(header.IPv4ProtocolNumber)
+
+	tuple.dst.addr = netHeader.DestinationAddress()
+	tuple.dst.port = tcpHeader.DestinationPort()
+
+	// TODO(gvisor.dev/issue/170): Need to support other transport protocols.
+	tuple.dst.protocol = uint8(header.TCPProtocolNumber)
+	tuple.dst.direction = CT_DIR_ORIGINAL
+
+	return tuple, nil
+}
+
+// getConnTrack returns connection if exists and sets status of connection
+// to established if the tuple is reply of an existing connection.
+func getConnTrack(hash uint32, s *Stack) (*connTrack, *tcpip.Error) {
+	tupleHash, err := s.connTrackTable[hash]
+	if !err {
+		return nil, tcpip.ErrUnknownProtocol
+	}
+	if tupleHash.ct == nil {
+		panic("tupleHash has null connection tracking entry")
+	}
+
+	// If this is the reply of new connection, set the connection
+	// status as ESTABLISHED.
+	ct := tupleHash.ct
+	if ct.status == CT_NEW && tupleHash.tuple.dst.direction == CT_DIR_REPLY {
+		ct.status = CT_ESTABLISHED
+	}
+	return ct, nil
+}
+
+// getInvertTuple creates inverted tuple for the given tuple.
+func getInvertTuple(tuple connTrackTuple) connTrackTuple {
+	var invertTuple connTrackTuple
+	invertTuple.src.addr = tuple.dst.addr
+	invertTuple.src.port = tuple.dst.port
+	invertTuple.src.protocol = uint16(header.IPv4ProtocolNumber)
+	invertTuple.dst.addr = tuple.src.addr
+	invertTuple.dst.port = tuple.src.port
+	invertTuple.dst.protocol = uint8(tuple.dst.protocol)
+	invertTuple.dst.direction = CT_DIR_REPLY
+	return invertTuple
+}
+
+// makeNewConn creates new connection.
+func makeNewConn(tuple, invertTuple connTrackTuple) *connTrack {
+	var ct connTrack
+	ct.status = CT_NEW
+	ct.tupleHash[CT_DIR_ORIGINAL].tuple = tuple
+	ct.tupleHash[CT_DIR_ORIGINAL].ct = &ct
+	ct.tupleHash[CT_DIR_REPLY].tuple = invertTuple
+	ct.tupleHash[CT_DIR_REPLY].ct = &ct
+
+	return &ct
+}
+
+// getTupleHash returns hash of the tuple. The fields used for
+// generating hash are seed(generated once for stack), source address,
+// destination address, source port and destination ports.
+func getTupleHash(tuple connTrackTuple, s *Stack) uint32 {
+	h := jenkins.Sum32(s.Seed())
+	h.Write([]byte(tuple.src.addr))
+	h.Write([]byte(tuple.dst.addr))
+	portBuf := make([]byte, 2)
+	binary.LittleEndian.PutUint16(portBuf, tuple.src.port)
+	h.Write([]byte(portBuf))
+	portBuf = make([]byte, 2)
```

Comment:
Can we reuse portBuf instead of allocating another?

Reply:
Done.